Add first class component cache

reeganviljoen opened the pull request:
closes #234

### What are you trying to accomplish?

Add experimental fragment caching for ViewComponent that is true to the ViewComponent ethos, namely:

- caching remains component-local (no hidden global state required to work)
- caching that can easily be debugged with standard debugging tools like breakpoints

### What approach did you choose and why?

I chose to add an opt-in caching API:

- `ViewComponent::Cacheable` + a `cache_on` macro that declares which instance methods contribute to the cache key
- cache keys include component identity + template identity (variant/format) + declared dependencies (via `ActiveSupport::Cache.expand_cache_key`) + a component digest
- the component digest invalidates cached output when the component’s source changes (Ruby file, templates, i18n sidecars) and when rendered child ViewComponents change
- template dependency extraction compiles templates to Ruby and uses Prism to find `render` dependencies
  - ERB uses stdlib `ERB::Compiler` (no Temple)
  - Slim/Haml are loaded lazily and only when those handlers are used

### Anything you want to highlight for special attention from reviewers?

- This is intentionally opt-in (global opt-in via `require "view_component/fragment_caching"` or per-component `include ViewComponent::Cacheable`)
- Added integration coverage proving parent cache invalidates when a child component template changes
- `cache_on` supports private methods and uses Rails cache key expansion for dependency values
- Known limitation: partial/layout string dependencies are not currently included in the digest (workaround: bump `RAILS_CACHE_ID`/`RAILS_APP_VERSION` on deploy)
- Follow-ups can extend the API (procs, conditionals/options, broader dependency coverage)

Commit:
add component controlled cache

Commit:
add changelog

Commit:
fix cacahe implementatation to work with all methods

Commit:
fix cacahe implementatation to work with all methods

Commit:
fix lint

reeganviljoen commented:
@joelhawksley @Spone @camertron I would appreciate any thoughts on this.

joelhawksley reviewed (commented):
@reeganviljoen THANK YOU for taking a crack at building something functional to address the long-overdue need for caching in ViewComponent. 

To be honest with you, I have not used caching in Rails nearly enough to feel like I can properly assess this PR in terms of how it would integrate with the existing mental model of caching in Rails. I'm happy to learn that context, but I won't have the space to do so for at least a couple of weeks. 

@Spone @boardfish have y'all done much with Rails' caching? @reeganviljoen if you know of anyone else in the community who would be a good reviewer here, I'd be happy to have them.

reeganviljoen commented:
> @reeganviljoen THANK YOU for taking a crack at building something functional to address the long-overdue need for caching in ViewComponent. 
> 
> To be honest with you, I have not used caching in Rails nearly enough to feel like I can properly assess this PR in terms of how it would integrate with the existing mental model of caching in Rails. I'm happy to learn that context, but I won't have the space to do so for at least a couple of weeks. 
> 
> @Spone @boardfish have y'all done much with Rails' caching? @reeganviljoen if you know of anyone else in the community who would be a good reviewer here, I'd be happy to have them.

@joelhawksley I really aprreciate the honesty, I will see if I can find anyone with cache experience to take a crack at reviewing this ❤️

boardfish commented:
@joelhawksley Afraid it's not something I've really interacted with directly until now, sorry! I'd be coming from a similar place. Great to see @reeganviljoen working at it, though ✨

mikodagatan commented on lib/view_component/base.rb:
```diff
@@ -319,6 +331,15 @@ def __vc_render_in_block_provided?
 
+    def __vc_render_template(rendered_template)
+      # Avoid allocating new string when output_preamble and output_postamble are blank
+      if output_preamble.blank? && output_postamble.blank?
```
Are there any chances only one of them would be blank?

reeganviljoen replied:
yes they can be individually toggled

mikodagatan replied:
Not sure, if I'm contributing to this. If it's safe to have one of them blank (since we don't have a statement to cover this having only one of them blank), then isn't it also safe to do the else statement `safe_output_preamble + rendered_template + safe_output_postamble` if both of them are blank?

reeganviljoen replied:
don't worry to much about the established code, a look at the cache mechanism is where the bulk of the eyes are needed

mikodagatan commented on lib/view_component/base.rb:
```diff
@@ -495,6 +516,14 @@ def sidecar_files(extensions)
+      def cache_on(*args)
+        class_eval <<~RUBY, __FILE__, __LINE__ + 1
+          def __vc_cache_args
+            #{args}
```
I think it may be better to not just include the args. You could also include the class name of the component to the key of the cache. 

This way, you will not overwrite the cache for the first component with another component that has the same args

FirstComponent.new(foo:, bar:) 
SecondComponent.new(foo: bar:)

mikodagatan replied:
Also, would you like to add the `updated_at` field if the arguments are model records? It would ensure that the cache is invalidated when you update the associated records.

reeganviljoen replied:
@mikodagatan to get this across the line I will add this in the next pr with docs and other nice to haves

BlakeWilliams commented on lib/view_component/base.rb:
```diff
@@ -50,6 +50,17 @@ def config
+    # Compoents can have a cache key that is used to cache the rendered output.
+    #
+    # @return [String]
+    def cache_key
```
Can we separate this into a separate module, similar to how we implemented slots?

reeganviljoen replied:
on it

reeganviljoen replied:
done

BlakeWilliams commented on lib/view_component/base.rb:
```diff
@@ -50,6 +50,17 @@ def config
+    def cache_key
+      @vc_cache_key = if defined?(__vc_cache_args)
+        Digest::MD5.hexdigest(
+          __vc_cache_args.map { |method| send(method) }.join("-")
```
How does this differ from Rails caching? Should we leverage their caching infrastructure to keep parity behavior wise?

reeganviljoen replied:
@BlakeWilliams solid idea so I looked at https://api.rubyonrails.org/v6.0.6.1/classes/AbstractController/Caching.html#method-i-view_cache_dependencies
and implemented similar logic, only is the class_attribute is leaking across classes, do you have an idea why and how would you fix that

reeganviljoen replied:
I have fixed the leaking

BlakeWilliams commented on lib/view_component/base.rb:
```diff
@@ -117,11 +128,12 @@ def render_in(view_context, &block)
-        if output_preamble.blank? && output_postamble.blank?
-          rendered_template
+        if cache_key.present?
+          Rails.cache.fetch(@vc_cache_key) do
```
```suggestion
          Rails.cache.fetch(cache_key) do
```

Lets avoid relying on this ivar since it requires `cache_key` to be called first. We should access one or the other consistently.

BlakeWilliams commented on lib/view_component/base.rb:
```diff
@@ -50,6 +50,17 @@ def config
+    #
+    # @return [String]
+    def cache_key
+      @vc_cache_key = if defined?(__vc_cache_args)
```
```suggestion
      return @__vc_cache_key if defined?(@__vc_cache_key)
      @__vc_cache_key ||= if defined?(__vc_cache_args)
```

I think we want to memoize this?

BlakeWilliams commented on lib/view_component/base.rb:
```diff
@@ -495,6 +516,14 @@ def sidecar_files(extensions)
 
+      def cache_on(*args)
+        class_eval <<~RUBY, __FILE__, __LINE__ + 1
+          def __vc_cache_args
```
I'm not sure I understand why this is defining a method if we're accepting symbols as arguments.

could this be:

```ruby
def cache_on(*args)
  @__vc_cache_args ||= args
end
```

reeganviljoen replied:
@BlakeWilliams this has chnaged please take a look when you can 🙏

Commit:
yeah I know it aint working, I am tired however, taking a nother look tomorow

ekampp commented on lib/view_component/base.rb:
```diff
@@ -516,12 +510,16 @@ def sidecar_files(extensions)
+      end
+
+      def view_cache_dependencies
+        return unless __vc_cache_dependencies.any?
```
```suggestion
        return if __vc_cache_dependencies.none?
```

This has a lower cognitive load because it avoids having to flip the sentence in your read when you read any.

reeganviljoen replied:
I generally disagree with that in when using a single condition but when using multiple conditions I prefer an if statement

Commit:
fix cache

Commit:
fix lint

Commit:
fix

reeganviljoen commented on lib/view_component/base.rb:
```diff
@@ -50,6 +50,17 @@ def config
+    def cache_key
+      @vc_cache_key = if defined?(__vc_cache_args)
+        Digest::MD5.hexdigest(
+          __vc_cache_args.map { |method| send(method) }.join("-")
```
this has been replaced

Commit:
modulerize code

Commit:
more cleanup

ekampp commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+    #
+    # @private
+    def view_cache_dependencies
+      return unless __vc_cache_dependencies.present? && __vc_cache_dependencies.any?
```
Consider this for reducing cognitive load:

```suggestion
      return if __vc_cache_dependencies.blank? || __vc_cache_dependencies.none?
```

reeganviljoen replied:
I actually agree with this one, I generally disagree with using unless when using multiple conditions, thanks

ekampp commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+    def view_cache_dependencies
+      return unless __vc_cache_dependencies.present? && __vc_cache_dependencies.any?
+
+      __vc_cache_dependencies.map { |dep| send(dep) }.compact
```
```suggestion
      __vc_cache_dependencies.filter_map { |dep| send(dep) }
```

reeganviljoen committed:
Apply suggestions from code review

Co-authored-by: Emil Kampp <40206+ekampp@users.noreply.github.com>

reeganviljoen committed:
Update lib/view_component/cacheable.rb

Co-authored-by: Emil Kampp <40206+ekampp@users.noreply.github.com>

JWShuff commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+      __vc_cache_dependencies.push(*args)
+    end
+
+    def inherited(child)
```
We took this for a spin in a few cases of our cached partials and view components that we are familiar with and largely it works for the base cases. This is working to correctly bust a cached VC when it changes or things in the render path 'above' (partials or VCs) it change.

What we aren't seeing is when a VC renders another VC or partial as a child that downstream changes of the cached VC pick up changes and handle it. I'm going to fork your branch and offer some test cases based on an approach we use in the [vc fragment caching gem](https://github.com/patrickarnett/view_component-fragment_caching) (what we're currently solving this with)

Effectively we want some way for a child partial or VC change (in either the `.rb` or template) to bust the cached parent.

reeganviljoen replied:
@JWShuff like touch: true in rails ?

reeganviljoen replied:
also big thanks for testing it, I can look at adding the touch true thing later this week

joelhawksley replied:
@reeganviljoen I'd expect us to test this with changes to child partials.

JWShuff reviewed (commented):
Reegan, really excited to see this. I'm going to take a shot at adapting our gem spec suite to your fork and open a PR at your fork with those specs by way of giving us a place to start that gives us the most cache functionality we can get.

JWShuff reviewed (commented):
Added a hypothetical approach to an integration test for the behavior.

reeganviljoen committed:
fix alphebtization

reeganviljoen committed:
add cache suhggestions

reeganviljoen commented:
@JWShuff I investigated a few of your suggestions, added a few, thanks, and others I am struggling with
luckily I think these aren't required for the initial feature and are more nice to haves that can be implemented in another pr(child cache busting)

reeganviljoen commented:
@joelhawksley with a live use case tested how do you feel about this now

reeganviljoen committed:
fix legacy ruby specs

reeganviljoen commented:
@joelhawksley, with this being tested @JWShuff, how do you feel about going forward with this?

joelhawksley commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+      __vc_cache_dependencies.filter_map { |dep| send(dep) }
+    end
+
+    # For handeling the output_preamble and output_postamble
```
```suggestion
    # For handling the output_preamble and output_postamble
```

joelhawksley commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+    # For handeling the output_preamble and output_postamble
+    #
+    # @private
+    def __vc_render_template(rendered_template)
```
IMO we should leave this inside Base as we'll be including this module manually to start.

reeganviljoen replied:
done, thank you

joelhawksley commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+      end
+    end
+
+    # For determing if a template is rendered with cache or not
```
```suggestion
    # Render component from cache if possible
```

joelhawksley commented on test/sandbox/test/rendering_test.rb:
```diff
@@ -1240,4 +1240,21 @@ def test_request_param
+    render_inline(CacheComponent.new(foo: "foo", bar: "baz"))
+
+    refute_selector(".cache-component__cache-key", text: component.view_cache_dependencies)
+    refute_selector(".cache-component__cache-message", text: "foo bar")
```
Can we add an assertion for what _should_ be shown here?

reeganviljoen replied:
done, thank you

joelhawksley reviewed (commented):
@reeganviljoen thanks for putting this together and testing it with @JWShuff. I'm interested in landing it as an optional, experimental module so we can make breaking changes outside of SemVer.

Besides the comments below, might you add some documentation? I'm happy to edit whatever you come up with ❤️

reeganviljoen committed:
Apply suggestions from code review

Co-authored-by: Joel Hawksley <joelhawksley@github.com>

reeganviljoen committed:
code review feedback

reeganviljoen commented on lib/view_component/base.rb:
```diff
@@ -29,6 +30,7 @@ def config
       end
     end
 
+    include ViewComponent::Cacheable
```
done

reeganviljoen committed:
make module fully optional;

reeganviljoen committed:
fix specs

reeganviljoen committed:
fix lint

reeganviljoen committed:
fix coberage

reeganviljoen committed:
add inherited component test

reeganviljoen committed:
fix tests

reeganviljoen committed:
merge inherited values

reeganviljoen committed:
fix tests

reeganviljoen committed:
fix lint

reeganviljoen commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,56 @@
+    # @private
+    def __vc_render_cacheable(rendered_template)
+      if view_cache_dependencies.present?
+        Rails.cache.fetch(view_cache_dependencies) do
```
@joelhawksley I don't really understand your question, the cache doesn't really care about how the component renders its content(i.e call, vs sidecar vs normal template vs inline template), it just cares about if the parameters marked for cache are changed

reeganviljoen replied:
@joelhawksley I have added the variant and format to the cache digest to make sure that it is more unique and that a change to either would generate a unique cache key

tmaier replied:
When I read the code correctly, then a component with the same cache dependencies and maybe the same variant and format would lead to a collusion in the rails cache.

This is why I would recommend to also add the `identifier`.

reeganviljoen replied:
@tmaier thanks for that I have added `identifier` to :__vc_cache_dependencies

reeganviljoen committed:
add polish

reeganviljoen committed:
add wip docs

reeganviljoen committed:
fix tests

reeganviljoen committed:
fix lint

reeganviljoen committed:
fix coverage

reeganviljoen committed:
fix lint

reeganviljoen committed:
fix lint

reeganviljoen committed:
fix missing coverage

reeganviljoen committed:
add format and varaiant to cache_digest

reeganviljoen committed:
add format and varaiant to cache_digest

reeganviljoen committed:
fix coverage

tmaier commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,44 @@
+    def view_cache_dependencies
+      return if __vc_cache_dependencies.blank? || __vc_cache_dependencies.none? || __vc_cache_dependencies.nil?
+
+      __vc_cache_dependencies.filter_map { |dep| send(dep) }.join("-")
```
Could this use a logic similar to what rails does? This would work then also if the cache dependencies would have a more complex object than a string.

```ruby
        def retrieve_cache_key(key)
          case
          when key.respond_to?(:cache_key_with_version) then key.cache_key_with_version
          when key.respond_to?(:cache_key)              then key.cache_key
          when key.is_a?(Array)                         then key.map { |element| retrieve_cache_key(element) }.to_param
          when key.respond_to?(:to_a)                   then retrieve_cache_key(key.to_a)
          else                                               key.to_param
          end.to_s
        end
```

Source: https://github.com/rails/rails/blob/0367cbf265ff25546230d2f668aa887e09995509/activesupport/lib/active_support/cache.rb#L125-L133

reeganviljoen replied:
Firstly thank you :heart: , I really appreciate the help on this, secondly I think that might be a good idea, so I will look into it and hopefully put something together and let you know

reeganviljoen committed:
add retrive ccache key to be consistent with rails

reeganviljoen committed:
Fix linting

reeganviljoen committed:
fix changelog stuff

reeganviljoen committed:
refactor cache logic

reeganviljoen committed:
add identifier

reeganviljoen committed:
compuye cache keys

reeganviljoen committed:
fix lint

tmaier commented on lib/view_component/cacheable.rb:
```diff
@@ -0,0 +1,57 @@
+  extend ActiveSupport::Concern
+
+  included do
+    class_attribute :__vc_cache_dependencies, default: [:format, :__vc_format, :identifier]
```
Should this be a Set so that we have automatic deduplication? This would allow that one calls `cache_on :current_locale` in the `ApplicationComponent`. Subsequent calls to `cache_on` would not lead to duplicates in `__vc_cache_dependencies`.

Side-note: `cache_on` with lambda support would also be interesting. E.g. `cache_on -> lambda { I18n.current_locale }`

tmaier commented on lib/view_component/base.rb:
```diff
@@ -293,6 +304,13 @@ def __vc_request
       @__vc_request ||= controller.request if controller.respond_to?(:request)
     end
 
+    # Fo use in caching
```
Spelling error

tmaier commented:
I have two questions about cache invalidation. Please understand, that I am coming from the perspective of [Ruby on Rails Managing Dependencies](https://guides.rubyonrails.org/caching_with_rails.html#managing-dependencies)
If the following will not be supported intentionally, then it should be documented in the guide.

## 1.

Rails takes the digest of a template file into account when considering to invalidate a cached object.
When I understand it correctly, this is handled by `ActionView::Digestor` in Rails.

Translated to ViewComponents, one would need to take the `.rb` file, `.html.erb` file and the `.yml` file into account. Essentially, all sidecar files. Is this something already now supported? I guess not.
If this is by purpose, we could document a workaround to invalidate the cache with each deploy by changing the environment variable `RAILS_CACHE_ID`.

## 2.

If we would have (1), the next step would be, that dependencies would invalidate the cache of a component.

Assume you would have `OneComponent` with this template:

```erb
OneComponent
<%= render TwoComponent.new %>
```

Then a change to `TwoComponent` should lead to an invalidation of the cache of `OneComponent`.
Was this what @JWShuff referred to in his comments?
I think a work-around would be the pattern described in <https://guides.rubyonrails.org/caching_with_rails.html#external-dependencies>

reeganviljoen commented:
> I have two questions about cache invalidation. Please understand, that I am coming from the perspective of [Ruby on Rails Managing Dependencies](https://guides.rubyonrails.org/caching_with_rails.html#managing-dependencies) If the following will not be supported intentionally, then it should be documented in the guide.
> 
> ## 1.
> Rails takes the digest of a template file into account when considering to invalidate a cached object. When I understand it correctly, this is handled by `ActionView::Digestor` in Rails.
> 
> Translated to ViewComponents, one would need to take the `.rb` file, `.html.erb` file and the `.yml` file into account. Essentially, all sidecar files. Is this something already now supported? I guess not. If this is by purpose, we could document a workaround to invalidate the cache with each deploy by changing the environment variable `RAILS_CACHE_ID`.
> 
> ## 2.
> If we would have (1), the next step would be, that dependencies would invalidate the cache of a component.
> 
> Assume you would have `OneComponent` with this template:
> 
> ```
> OneComponent
> <%= render TwoComponent.new %>
> ```
> 
> Then a change to `TwoComponent` should lead to an invalidation of the cache of `OneComponent`. Was this what @JWShuff referred to in his comments? I think a work-around would be the pattern described in https://guides.rubyonrails.org/caching_with_rails.html#external-dependencies

after looking at this it seems we may need to build a `ViewComponent::Digestor` to do the same thing but for view component but I'll first need to dig a bit deeper on how rails does this

reeganviljoen committed:
refactor

reeganviljoen committed:
add set

reeganviljoen committed:
Add cache refistry and alighn cache with how action view does it

reeganviljoen committed:
namespace registry

reeganviljoen committed:
add magic comment

reeganviljoen committed:
fix failing rails 6 specs

reeganviljoen committed:
Add the start of an actual digestor

reeganviljoen committed:
add template digetor that usses an ast

reeganviljoen commented:
@tmaier I have built a cache digetor that makes use of temple and prism to build an ast(as apposed to rails use of regexes) to build a digest, would you mind taking a look :pray:

reeganviljoen committed:
refactor digetor a bit

reeganviljoen committed:
fix indentation

reeganviljoen committed:
refactor

reeganviljoen committed:
get pr up top date

reeganviljoen committed:
fix merge issue

reeganviljoen committed:
try get tests to pass

reeganviljoen committed:
try get tests to pass

reeganviljoen committed:
fix rails 8 test

reeganviljoen committed:
fix soem artifcats

reeganviljoen committed:
fix test

reeganviljoen committed:
make primer pass

reeganviljoen committed:
fix linting

reeganviljoen committed:
fix linting

reeganviljoen committed:
fix alloactor spec

reeganviljoen committed:
fix tests

reeganviljoen committed:
make primer pass

reeganviljoen committed:
add inline erb cache component test

joelhawksley committed:
Merge branch 'main' into rv_add_component_caching

Commit:
Merge branch 'main' into rv_add_component_caching

Commit:
fix ci

Commit:
fix primer ci failures by making ast lazy load dependecies

Commit:
treat `temple`, `slim`, `haml` as optional dependencies (no boot-time hard requires), make parent component cache key changes when a rendered child component changes.

Commit:
fix ci

Commit:
Refine fragment caching key + invalidation

Commit:
Normalize cache_on dependencies and memoize digests

Commit:
update docs

Commit:
refactor dependecy extractor

Commit:
Optimize digest and template dependency extraction

Avoid Temple for ERB by compiling templates to Ruby, and reduce allocations in digest/dependency extraction to keep caching overhead low.

reeganviljoen committed:
Merge branch 'main' into rv_add_component_caching

joelhawksley committed:
Merge branch 'main' into rv_add_component_caching

joelhawksley commented on docs/CHANGELOG.md:
```diff
@@ -10,6 +10,10 @@ nav_order: 6
 
 ## main
 
+* Add experimental support for caching.
```
```suggestion
* Add experimental support for caching by including `ViewComponent::ExperimentallyCacheable`.
```

joelhawksley commented on lib/view_component/base.rb:
```diff
@@ -49,12 +49,13 @@ def config
     include ViewComponent::Translatable
     include ViewComponent::WithContentHelper
 
+    class_attribute :__vc_strip_trailing_whitespace, instance_accessor: false, instance_predicate: false, default: false
```
Why was this moved? The associated comment was not moved.

joelhawksley commented on lib/view_component/cache_digestor.rb:
```diff
@@ -0,0 +1,91 @@
+        sources.concat(dependency_digests(dependencies))
+      end
+
+      component_class.sidecar_files(["yml"]).sort.each do |path|
```
Is `yaml` also valid?

reeganviljoen replied:
it is

joelhawksley commented on lib/view_component/prism_render_dependency_extractor.rb:
```diff
@@ -0,0 +1,54 @@
+require "prism"
+
+module ViewComponent
+  class PrismRenderDependencyExtractor
```
This is very similar to @jhawthorn's https://github.com/jhawthorn/actionview_precompiler/blob/main/lib/actionview_precompiler/render_parser.rb, perhaps we could reuse it instead of writing our own?

Also relevant: https://github.com/rails/rails/pull/50944 and https://github.com/rails/rails/pull/42458.

joelhawksley commented on test/sandbox/test/rendering_test.rb:
```diff
@@ -1339,6 +1339,90 @@ def test_around_render
   end
 
+  def test_inline_cache_component
+    return if Rails.version < "7.0"
```
We no longer support < `7.0` so let's remove this and the other early returns in this file.

joelhawksley reviewed (commented):
Overall, this is in great shape. Lots of thoughts to consider before we land this but I think we're on the right track towards having something I'd be comfortable landing ❤️

joelhawksley commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,50 @@
+- `cache_on` accepts method names; the returned values are expanded via `ActiveSupport::Cache.expand_cache_key` (so Active Record models, `GlobalID`, arrays, and plain strings work as expected).
+- Methods listed in `cache_on` may be private (the cache dependency reader calls `send`).
+- The cache key includes a digest of the component source (Ruby + templates + i18n sidecars) and rendered child ViewComponents.
+- Partial/layout string dependencies are not currently included in the digest; use `RAILS_CACHE_ID`/`RAILS_APP_VERSION` if you need to invalidate on deploy.
```
Also, I think we should have a test that captures this shortcoming so we can fix it eventually.

Commit:
code review easy wins

Commit:
code review: added a dedicated regression test for child partial dependency changes

Commit:
code review: added shared integration-style spec approach from fork for potential adoption

Commit:
code review: consider reusing parser approach from existing project instead of custom implementation.

Commit:
add cacing benchmark

reeganviljoen committed:
Update docs/guide/caching.md

Co-authored-by: Joel Hawksley <joelhawksley@github.com>

reeganviljoen commented on docs/guide/caching.md:
````diff
@@ -0,0 +1,50 @@
+
+```ruby
+class CacheComponent < ViewComponent::Base
+  include ViewComponent::Cacheable
````
done thanks

Commit:
fix code that was moved incorectly

reeganviljoen commented on lib/view_component/cache_digestor.rb:
```diff
@@ -0,0 +1,91 @@
+      sources = []
+
+      identifier = component_class.identifier
+      sources << file_contents(identifier) if identifier
```
sorry I this was something I forgot to remove while debugging

reeganviljoen commented on lib/view_component/fragment_caching.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
```
I see and that is a valid point so I removed it

Commit:
imrpove performnce

Commit:
add refactors to improve benchmarks

reeganviljoen commented:
After a few refactors to improve the benchmark numbers it now looks a lot better 
```
ruby 4.0.0 (2025-12-25 revision 553f1675f3) +YJIT +PRISM [arm64-darwin24]
Warming up --------------------------------------
       non_cacheable    51.748k i/100ms
      cacheable_miss    41.394k i/100ms
       cacheable_hit    39.364k i/100ms
Calculating -------------------------------------
       non_cacheable    558.656k (± 2.3%) i/s    (1.79 μs/i) -     11.178M in  20.019916s
      cacheable_miss    491.207k (± 2.2%) i/s    (2.04 μs/i) -      9.852M in  20.067487s
       cacheable_hit    396.384k (± 4.7%) i/s    (2.52 μs/i) -      7.912M in  20.036403s

Comparison:
       non_cacheable:   558656.4 i/s
      cacheable_miss:   491206.7 i/s - 1.14x  slower
       cacheable_hit:   396383.8 i/s - 1.41x  slower
```

Commit:
make benchmark cmponents more expensive too really show ccache performnce

Commit:
Merge branch 'main' into rv_add_component_caching

Commit:
fix ci

Commit:
fix ci

reeganviljoen commented on lib/view_component/cache_digestor.rb:
```diff
@@ -0,0 +1,91 @@
+
+      return "" if @visited_components.key?(component_class.name)
+
+      @visited_components[component_class.name] = true
```
this has been reworked now thank you

Commit:
add test for joels noted limitation

Commit:
fix tests

Commit:
fix ci

Commit:
use ActionviewPrecompiler instread of our own stuff

Commit:
add action view precompiler

Commit:
Merge branch 'main' into rv_add_component_caching

Commit:
fix tests

Commit:
fix accidental commits

Commit:
fix rails-main

Commit:
fix rails main

Commit:
fix ci

Commit:
fix ci

Commit:
fix rails main

Commit:
fix vale issues

joelhawksley commented on docs/guide/caching.md:
````diff
@@ -0,0 +1,45 @@
+<p><%= "#{foo} #{bar}" %></p>
+```
+
+`cache_on` accepts method names. Returned values are expanded via `ActiveSupport::Cache.expand_cache_key`, so Active Record models, `GlobalID`, arrays, and plain strings work as expected.
````
```suggestion
Returned values are expanded via `ActiveSupport::Cache.expand_cache_key`, so Active Record models, `GlobalID`, arrays, and plain strings work as expected.
```

joelhawksley commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,45 @@
+
+To enable caching, include `ViewComponent::ExperimentallyCacheable`.
+
+Components implement caching by marking dependencies using `cache_on`:
```
Is it true that one _must_ call `cache_on`? What if you don't?

joelhawksley reviewed (commented):
Getting closer! Thank you for your hard work ❤️

Commit:
Address component caching review feedback

reeganviljoen commented:
@joelhawksley and everyone who has reviewed or tested this so far, I have come back to this with a few fixes addressing the latest comments.

The main changes are:

- moved the API to `cache do ... end` and updated the docs around that
- kept caching opt-in per component while experimental
- switched the fragment reads/writes to go through Rails fragment caching APIs
- added `sidecar_templates` / `sidecar_translations`
- simplified the digestor and added `ViewComponent::CacheDigestor.digest`
- made cache hit/miss recording respect the instrumentation setting
- added/precomputed the component digest when template caching is enabled
- added coverage for ERB, Slim, Haml, and Jbuilder template handlers
- kept the partial/layout string dependency limitation documented and covered by tests

I think this now addresses the comments raised here, but I would really appreciate another pass before we try to land it.

@JWShuff @tmaier if either of you have time to test this against real cached component use cases again, that would be especially helpful.

Thanks again everyone for all the review and patience on this.

Commit:
Merge branch 'main' of https://github.com/ViewComponent/view_component into rv_add_component_caching

# Conflicts:
#	.tool-versions
#	Gemfile.lock
#	docs/CHANGELOG.md
#	gemfiles/rails_main.gemfile.lock
#	test/sandbox/test/rendering_test.rb

Commit:
Fix CI failures after merge

Commit:
Refactor lint-sensitive chains

Commit:
Handle Ruby head ERB line offsets

Commit:
Keep CI fix scoped

Commit:
Keep component caching isolated

Commit:
Stabilize rendering allocation check

reeganviljoen committed:
Merge branch 'main' into rv_add_component_caching

Commit:
Merge remote-tracking branch 'upstream/main' into rv_add_component_caching

Commit:
Align Standard pin with main

Commit:
Merge remote-tracking branch 'origin/rv_add_component_caching' into rv_add_component_caching

timburgan committed:
Fix cache_if with private methods

cache_if evaluated Symbol/String conditions with public_send, so the documented
example (cache_if :cacheable? with a private cacheable?) crashed at render with
NoMethodError. Use send, consistent with how the cache block itself is evaluated
via instance_exec. The existing cache_if test now uses a private method, and a
positive test covers a private condition that enables caching.

timburgan committed:
Centralize sidecar template/translation extension lists

The cache digestor reached into ActionView::Template.template_handler_extensions
and hardcoded %w[yml yaml], the latter duplicating Translatable's
TRANSLATION_EXTENSIONS. Expose sidecar_templates (Base) and sidecar_translations
(Translatable) so the handler-extension coupling lives in one auditable place and
the translation list can't drift from i18n loading. Addresses review feedback on
the PR (sidecar_templates/sidecar_translations).

timburgan committed:
Surface template AST build failures instead of swallowing them

The bare rescue turned any ActionView::Template/handler-call signature break into
a silent nil, which degrades to empty dependencies and stale caches. Capture the
error and emit template_ast_build_failed.view_component instrumentation, and add
per-handler canary tests (erb/slim/haml) so an upgrade break fails loudly in CI.

timburgan committed:
Feature-detect ActionView::Renderer#cache_hits (internal API)

cache_hits is :nodoc: and only annotates the render log; feature-detect it with
respond_to? so a future Rails change can't raise during caching. The supported
hit/miss signal remains the read_fragment/write_fragment ActiveSupport
notifications, which fire regardless.

timburgan committed:
Include I18n.locale in the cache key

The cache key was composed only of the component digest and the cache block's
dependencies, so a component with sidecar translations served the cached fragment
from whichever locale rendered first (e.g. English output for a French request).
Add I18n.locale to the key and cover it with a translatable, cacheable component
that must render the correct language per locale.

timburgan committed:
Remove redundant inherited hook that broke late cache declarations

__vc_cache_key_block/__vc_cache_if are class_attributes, which already inherit.
The extra inherited hook eagerly copied the parent's then-current value into each
subclass, so declaring cache on a parent AFTER a subclass was defined left the
subclass with a stale nil and it silently would not cache. Rely on class_attribute
inheritance instead.

timburgan committed:
Invalidate cache on changes to indirectly-rendered child components

Dependency extraction ran only over template source, so a child rendered from a
Ruby method (e.g. a helper called from the template) was not a tracked dependency
and changes to it did not bust the parent digest. Also scan the component's own
Ruby source for component renders (render Foo.new) and recurse, so indirect
children invalidate the parent like template-level renders do.

timburgan committed:
Add caching regression tests for edge cases and invalidation depth

Cover def-call components, ActiveRecord-like models (hit + invalidate on update),
child .rb changes, indirectly-rendered children, cyclic/self-referential digest
termination, variant isolation, HTML-safety round-trip, and cache-block
inheritance ordering.

timburgan committed:
Trim internal comments and mark sidecar helpers @private

Condense the rationale comments on the AST-build rescue, cache_hits guard, and
extract_component_renders to match the terse style of the surrounding internal
classes.

Mark sidecar_templates/sidecar_translations @private: both are only called
cross-class by the digestor and translation loader, so they stay Ruby-public but
out of the documented API surface.

joelhawksley commented:
@reeganviljoen when you have some time, I'd be curious about your thoughts on @timburgan's PR ❤️

reeganviljoen commented:
@joelhawksley I have left a small comment on @timburgen's pr, the preliminary review I did leads me to believe it would be a good course of action to bring his commits in here

Commit:
Reset component cache state between renders

Commit:
Merge upstream/main into rv_add_component_caching

Commit:
Merge remote-tracking branch 'upstream/main' into rv_add_component_caching

# Conflicts:
#	docs/CHANGELOG.md

joelhawksley commented:
Closing in favor of https://github.com/ViewComponent/view_component/pull/2685, which I just merged! If you can, please test the latest release in your apps and provide feedback ❤️
